monitoring/procfs.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2017 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/gravitational/trace"
+)
+
+// ProcNet* are standard paths to Linux procfs information on sockets
+const (
+	ProcNetTCP  = "/proc/net/tcp"
+	ProcNetTCP6 = "/proc/net/tcp6"
+	ProcNetUDP  = "/proc/net/udp"
+	ProcNetUDP6 = "/proc/net/udp6"
+	ProcNetUnix = "/proc/net/unix"
+)
+
+// SocketState stores Linux socket state
+type SocketState uint8
+
+// Possible Linux socket states
+const (
+	Established SocketState = 0x01
+	SynSent                 = 0x02
+	SynRecv                 = 0x03
+	FinWait1                = 0x04
+	FinWait2                = 0x05
+	TimeWait                = 0x06
+	Close                   = 0x07
+	CloseWait               = 0x08
+	LastAck                 = 0x09
+	Listen                  = 0x0A
+	Closing                 = 0x0B
+)
+
+// TCPSocket contains details on a TCP socket
+type TCPSocket struct {
+	LocalAddress  net.TCPAddr
+	RemoteAddress net.TCPAddr
+	State         SocketState
+	TXQueue       uint
+	RXQueue       uint
+	TimerState    uint
+	TimeToTimeout uint
+	Retransmit    uint
+	UID           uint
+	Inode         uint
+}
+
+// UDPSocket contains details on a UDP socket
+type UDPSocket struct {
+	LocalAddress  net.UDPAddr
+	RemoteAddress net.UDPAddr
+	State         SocketState
+	TXQueue       uint
+	RXQueue       uint
+	UID           uint
+	Inode         uint
+	RefCount      uint
+	Drops         uint
+}
+
+// UnixSocket contains details on a Unix socket
+type UnixSocket struct {
+	RefCount uint
+	Protocol uint
+	Flags    uint
+	Type     uint
+	State    uint
+	Inode    uint
+	Path     string
+}
+
+type socketparser func(string) (interface{}, error)
+
+func parseSocketFile(parse socketparser, fpath string) (interface{}, error) {
```

Comment:
No, I really did not mean this. I'm sorry for not making this clearer.
I just wanted to keep the file reading code common to all parsers:
```go
type addrParser func(line string) error

func parseSocketFile(path string, parse addParser) error {
	fp, err := os.Open(path)
	defer fp.Close()
	if err != nil {
		return trace.ConvertSystemError(err)
	}
	lineScanner := bufio.NewScanner(fp)
	lineScanner.Scan() // Drop header line
	for lineScanner.Scan() {
		err := parse(lineScanner.Text())
		if err != nil {
			return trace.Wrap(err)
		}
	}
	if err := lineScanner.Err(); err != nil {
		return trace.Wrap(err)
	}
	return nil
}
```
and then just use it to parse individual sockets:
```go
// GetTCPSockets reads specified file formatted as /proc/net/tcp{,6} and returns slice of *TCPSocket
func GetTCPSockets(path string) ([]*TCPSocket, error) {
	var sockets []TCPSocket
	err := parseSocketFile(path, func(line string) error {
		socket, err := parseTCPSocket(line)
		if err != nil {
			return trace.Wrap(err)
		}
		sockets = append(sockets, socket)
		return nil
	})
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return sockets, nil
}

// newTCPSocketFromLine parses line in /proc/net/tcp{,6} format and returns TCPSocket object or an error
func newTCPSocketFromLine(line string) (*TCPSocket, error) {
	// sl  local_address rem_address   st tx_queue rx_queue tr tm->when retrnsmt   uid  timeout inode
	//  0: 00000000:0035 00000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 18616 1 ffff91e759d47080 100 0 0 10 0
	// reference: https://github.com/ecki/net-tools/blob/master/netstat.c#L1070
	var (
		sl         int
		localip    []byte
		remoteip   []byte
		tr         int
		tmwhen     int
		retransmit int
		timeout    int
		tails      string
	)
	tcpSocket := &TCPSocket{}
	_, err := fmt.Sscanf(line, "%d: %32X:%4X %32X:%4X %2X %8X:%8X %2X:%8X %8X %d %d %d %1s",
		&sl, &localip, &tcpSocket.LocalAddress.Port, &remoteip, &tcpSocket.RemoteAddress.Port,
		&tcpSocket.State, &tcpSocket.TXQueue, &tcpSocket.RXQueue, &tr, &tmwhen, &retransmit,
		&tcpSocket.UID, &timeout, &tcpSocket.Inode, &tails)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	tcpSocket.LocalAddress.IP = hexToIP(localip)
	tcpSocket.RemoteAddress.IP = hexToIP(remoteip)
	return tcpSocket, nil
}
````